File: 0215-kth-largest-element-in-an-array/0215-kth-largest-element-in-an-array.cpp
class Solution {
public:

    int findKthLargest(vector<int>& nums, int k) {
        return quickselect(nums, k);
    }

    int quickselect(vector<int>& nums, int k) {
      int pivot = nums[rand() % nums.size()];

      vector<int> left;
      vector<int> mid;
      vector<int> right;

      for (int num : nums) {
        if (num > pivot) {
          left.push_back(num);
        } else if (num < pivot) {
          right.push_back(num);
        } else {
          mid.push_back(num);
        }
      }

      if (k <= left.size()) {
        return quickselect(left, k);
      }

      if (left.size() + mid.size() < k) {
        return quickselect(right, k - left.size() - mid.size());
      }

      return pivot;
    }


};

File: 0215-kth-largest-element-in-an-array/0215-kth-largest-element-in-an-array.py
from heapq import heappush, heappop
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:

        # def quickSelect(left, right, k):
        #     if left == right:
        #         return

        #     pivotIdx = random.randint(left, right)
        #     pivotVal = nums[pivotIdx]

        #     i, j, p = left, left, right # [left:i] > val, [i:p + 1] == val, [p + 1] < val

        #     while j <= p:
        #         if nums[j] > pivotVal:
        #             nums[i], nums[j] = nums[j], nums[i]
        #             i += 1
        #             j += 1
        #         elif nums[j] < pivotVal:
        #             nums[p], nums[j] = nums[j], nums[p]
        #             p -= 1
        #         else:
        #             j += 1

        #     if k <= i - left:
        #         quickSelect(left, i - 1, k)
        #     elif k <= p - left + 1:
        #         return
        #     else:
        #         quickSelect(p + 1, right, k - (p - left + 1))
            
        # quickSelect(0, len(nums) - 1, k)

        # return nums[k - 1]

        minHeap = []

        for num in nums:
            heappush(minHeap, num)

            if len(minHeap) > k:
                heappop(minHeap)

        return minHeap[0]
            
       

        

